### ISSUE TITLE
Interrupting omni completion may insert junk

### ISSUE THREAD

**Issue by seandewar** [prabirshrestha/vim-lsp#1087]
With `set omnifunc=lsp#complete`, trigger omni-completion:

![image](https://user-images.githubusercontent.com/6256228/110531767-e453f000-8113-11eb-9bcf-be2a73826a35.png)

Then, interrupt the completion with a weird key like `<C-g>` or `<C-v>`:

![image](https://user-images.githubusercontent.com/6256228/110531784-ea49d100-8113-11eb-9a2a-4d70b7536807.png)

This appears to trigger for some keys that begin a multi-key sequence insert mode mapping.

The issue seems to be caused by [this use of `feedkeys()`](https://github.com/prabirshrestha/vim-lsp/blame/0c77beb2eb99263e69a25b4d66e7ca2cd15cb0ef/autoload/lsp/ui/vim/completion.vim#L57).
Happens in Vim and Neovim, also with `-u NORC`.

**Comment by rhysd** [comment]
@prabirshrestha @mattn @hrsh7th I think the following line causes this issue.

https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/ui/vim/completion.vim#L57

Would you tell me why `feedkeys()` is necessary here? For example, doesn't replacing it with simple `:call` work?

```diff
diff --git a/autoload/lsp/ui/vim/completion.vim b/autoload/lsp/ui/vim/completion.vim
index dbfe13e..25bda10 100644
--- a/autoload/lsp/ui/vim/completion.vim
+++ b/autoload/lsp/ui/vim/completion.vim
@@ -54,7 +54,7 @@ function! s:on_complete_done() abort
   let s:context['complete_position'] = l:managed_user_data['complete_position']
   let s:context['server_name'] = l:managed_user_data['server_name']
   let s:context['completion_item'] = l:managed_user_data['completion_item']
-  call feedkeys(printf("\<C-r>=<SNR>%d_on_complete_done_after()\<CR>", s:SID()), 'n')
+  call s:on_complete_done_after()
 endfunction
 
 "
```

**Comment by hrsh7th** [comment]
We need `feedkeys` to detect `commit character` for now.

The `CompleteDone` fires before inserting `commit character` so we should delay it until inserting.

We have multiple options, I think.

1. Still use `feedkeys` (may insert junk, but no flicker)
2. Use `timer_start` (may flicker, but can be avoidable some problems)
3. Provide keymapping to confirm completion item.

**Comment by rhysd** [comment]
Thank you for the detailed explanation. I understood intention. (And I need to learn about commit characters.)

I thought 2. was good choice but I should try it locally at first.

**Comment by casr** [comment]
Thanks for reopening this @mattn. I've noticed this too that when I have:
```
set omnifunc=lsp#complete
```

I can trigger random junk to appear with typing regular characters. This is the result of me typing `oset `:

https://user-images.githubusercontent.com/566719/130405567-c5af8cee-1437-435d-977a-519f250f2b0e.mov

(This is a `.vimrc` file using `vim-language-server` but I've noticed this in JavaScript/TypeScript files too)

**Comment by hrsh7th** [comment]
@casr I think you set `longuest` to your `completeopt`.
Could you check the output of `:echomsg &completeopt`?

**Comment by casr** [comment]
```
:echomsg &completeopt
menu,preview
```

**Comment by mattn** [comment]
Is this still happen?

**Comment by casr** [comment]
> Is this still happen?

I tried to reproduce this bug with the version of the plugin that was available at the time it was opened but I was unable to. I was hoping to contrast it with the latest version to see if it was fixed for sure. In either case the bug doesn't present for me anymore.

Things that changed between then and now for me is a newer, faster computer and a different Vim/Neovim version. Sorry I can't be more helpful!
